### ISSUE TITLE
Signing verification of new signature files

### ISSUE THREAD

**Issue by cbridgha** [OpenLiberty/docs#4842]
Very soon - we will be producing *.sig (signature) files for each package of a release.

For example:  openliberty-21.0.0.10.zip.sig. (it will always just append a .sig extension to existing filename).

We need a simple doc page explaining what the files are, and how they can be verified using a downloadable public key that will be provided (on IBM's DHE download site) - so a link to this public key is required on this page.

Instructions on how to use an openssl command that is available on every OS we support.

An example command: `openssl dgst -sha256 -verify WebSphereLiberty_06-02-2021.pem -signature openliberty-javaee8-21.0.0.8.zip.sig openliberty-javaee8-21.0.0.8.zip`

**Comment by chirp1** [comment]
- Include a blurb in the WL IBM Docs and a link to the OL topic.
- The TWAS version of this issue is at https://github.ibm.com/websphere/twas-docs/issues/4910. Have the same person write the OL and tWAS topics.

**Comment by chirp1** [comment]
The epic isn't going into 21.0.0.12 as the design for the epic wasn't approved for the release. Removed 21.0.0.12 label.

**Comment by cbridgha** [comment]
Here is a rough draft for content:

All Open Liberty downloadable zips and archives are signed with the Liberty [public key](https://public.dhe.ibm.com/ibmdl/export/pub/software/openliberty/sign/public_keys/WebSphereLiberty_06-02-2021.pem). 

To verify, user must first download the [public key](https://public.dhe.ibm.com/ibmdl/export/pub/software/openliberty/sign/public_keys/WebSphereLiberty_06-02-2021.pem) and along with the downloaded zip package, also download the associated *.sig file

An example command using the command OpenSSL:

```
> openssl dgst -sha256 -verify WebSphereLiberty_06-02-2021.pem -signature openliberty-javaee8-21.0.0.8.zip.sig openliberty-javaee8-21.0.0.8.zip
```

Should see:  `Verified OK`

**Comment by cbridgha** [comment]
In addition - we should have a "sub" topic for verifying the Liberty packages up on Maven Central...

To verify those packages use the following instructions:

//GPG signing Verification
// download public key to local machine
`gpg2 --keyserver hkp://keyserver.ubuntu.com --recv-keys 46102B8E`
(`46102B8E` is the key id for the WebSphere Liberty public key)

// Editing trust of key owner
```
gpg2 --edit-key "WebSphere Liberty"
   gpg> trust
```
      Set trust 1-5(should be 5)


// verify file
`gpg2 --verify openliberty-runtime-21.0.0.7.zip.asc`

**Comment by chirp1** [comment]
Purpose: Customers download public key and use signature file to verify the IBM signature.

**Comment by dmuelle** [comment]
OpenSSL dgst doc: https://www.openssl.org/docs/man1.1.1/man1/openssl-dgst.html

**Comment by ramkumar-k-9286** [comment]
Hi Chuck @cbridgha 

Wanted your opinion on something: 

Where would this new document fit in the **Table Of Contents**:

https://docs-draft-openlibertyio.mybluemix.net/docs/22.0.0.1/overview.html

 Would it come under **Basics** / **Installing Open Liberty beta releases** /  ?

Regards,
Ramkumar

**Comment by dmuelle** [comment]
To clarify, I think two most likely places for the topic in the docs would be as a top-level entry in either the `Basics` section or the `Security` section. I'm thinking this is probably closer to Basics than security but I could see it either way

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

The following is the draft content ( before putting it into any document)
____

Signature files (*.sig) are produced for every package of an Open Liberty release.  You can verify these signature files by using the corresponding public key that is provided on IBM’s DHE download site.

For example, IBM uses its private key to digitally sign the latest Open Liberty release. You can use the Open Liberty public key that is provided on IBM’s DHE download site to check the signature. Verify that the Open Liberty release package was released by IBM and that it has not been modified since its release.

To verify, you must first download the [public key](https://public.dhe.ibm.com/ibmdl/export/pub/software/openliberty/sign/public_keys/WebSphereLiberty_06-02-2021.pem) along with the downloaded *.zip package, also download the associated *.sig file.

An example that uses the [OpenSSL command](https://www.openssl.org/docs/man1.1.1/man1/openssl-dgst.html):
```
> openssl dgst -sha256 -verify WebSphereLiberty_06-02-2021.pem -signature openliberty-javaee8-21.0.0.8.zip.sig openliberty-javaee8-21.0.0.8.zip
```

Should see:  

`Verified OK`

**Verify Liberty packages on Maven Central**

To verify the packages use the following instructions:
```
//GPG signing Verification
// download public key to local machine

>gpg2 --keyserver hkp://keyserver.ubuntu.com --recv-keys 46102B8E
(46102B8E is the key id for the WebSphere Liberty public key)

// Editing trust of key owner
gpg2 --edit-key "WebSphere Liberty"
   gpg> trust


      Set trust 1-5(should be 5)
// verify file

gpg2 --verify openliberty-runtime-21.0.0.7.zip.asc
```

**Comment by dmuelle** [comment]
a few comments-

- we need to make sure references to IBM are handled properly. This looks ok as far as I know but we'll want to run it by Alasdair when the draft is ready. I'm not sure if we want to claim that OL packages are "released by IBM".
- `You can use the Open Liberty public key that is provided on IBM’s DHE download site to check the signature. Verify that the Open Liberty release package was released by IBM and that it has not been modified since its release.`
--->
`You can use the Open Liberty public key to check the signature, verify that the Open Liberty release package was released by IBM and that it was not modified since its release.`
- `To verify, you must first download the public key along with the downloaded *.zip package, also download the associated *.sig file.`
This is as run-on sentence, also says to ":download... the downloaded zip file..." which doesn't make sense. Maybe
```
To verify an Open Liberty release, you must download the following files.
- the public key  
- An Open Liberty *.zip package
- the *.sig file for the release package that you want to verify.
```

Are the sig files a separate download or are they included in the package download? Are sig files available for any OL release version?


- `An example that uses the OpenSSL command:`
`After you download the files, you can run the following OpenSSL command from the command line to verify the package. 

[example]

This example uses the `WebSphereLiberty_06-02-2021.pem` public key file and `openliberty-javaee8-21.0.0.8.zip.sig` signature file to verify the `openliberty-javaee8-21.0.0.8.zip` release package. Replace these values according to the package that you want to verify. If the verification is successful, the command produces the following console output.

```
Verified OK
```


### Verify Liberty packages on Maven Central

 this section doesnt provide enough context to make sense unless you already know exactly what you doing. What is this code? Where do you run it? are these commands? How do you set the trust? Dont tell users to set it 1-5 if it should be 5, just tell them to set it to 5. But it's not clear how to do that or what it even means.
 
Looks like `gpg2` is a Linux encryption tool- is the process the same on Windows? 

Public Key link provided ? - does it change ?

**Comment by ramkumar-k-9286** [comment]
SUGGESTED FILE NAME:

`verifying-package-signatures.adoc`

**Comment by dmuelle** [comment]
Hi @cbridgha - we are working on this draft but need a little more information to complete it. Can you get back to us on the following questions?

- Are the sig files a separate download or are they included in the OL .zip package download?
    <cdb>Yes they will always appear alongside the original zip files. (*.zip.sig)</cdb>
- Are sig files available for any OL release version? How far back does support for them go?
    <cdb>They will first start appearing in 22.0.0.1 release and on...  </cdb>
- The examples here use the `WebSphereLiberty_06-02-2021.pem` file, as dos the provided DHE link. Since this file is dated for 2021, how long is it valid? Is there a stable link where users can go to get whatever the most current version of the key file is? <cdb>. This is the public key, and the date was when the certificate was "born" - in general its good for 2 years, and when we need to refresh it, there will be another key file with a later date</cdb>

**Verifying Packages on Maven**
Can you summarize what is happening in this example? The code comments don't provide much context. 

- What does  the `gpg2` command have to do with Maven?
- What is the "trust" in this context and how do you set it?
- "46102B8E is the key id for the WebSphere Liberty public key" - is that a stable value that is always the same?
   <cdb>. I'll add a link from Sonotype(maven central) that may give better context - at least around generating the signatures. https://central.sonatype.org/publish/requirements/gpg/  .
Here is a link on setting the "trust" of a certificate - we would tell customers to set it to 4 or 5 - we are telling them the public key is not compromised - they downloaded(https://www.gnupg.org/gph/en/manual/x334.html)
Yes the shortID of the key will not change as long as the public key doesn't expire(2 years).   When a new key is needed, we will change the instructions.</cdb>

Thanks

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

Please find the draft site link for the doc:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

Will need to incorporate info on Maven as well as other queries shared with Chuck.

Regards,
Ramkumar

**Comment by cbridgha** [comment]
Updated David's comment above...

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Please find the draft site link for the doc:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

Incorporated info on Maven

Regards,
Ramkumar

**Comment by dmuelle** [comment]
@ramkumar-k-9286 - the info for maven iosn't showing up in the draft.

comments for the first section:

> For example, IBM uses its private key to digitally sign the latest Open Liberty release. You can use the Open Liberty public key to check the signature, verify that the Open Liberty release package was released by IBM and that it was not modified since its release.

I think you can remove "For example," here. It isn't really an example so much as a description of the process . Also needs a comma after "...released b y IBM,". I'm still not sure that's the wording we want for OL but leave as is for now.

The bulleted list of files  is not rendering- prob needs a blank line inserted before list

For the release packages, Chuck says "They will first start appearing in 22.0.0.1 release and on.." so I think we should use 22.0.0.1 as the release version in all the examples. These examples use 21.0.0.8. Not sure if we're providing files for previous releases

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Corrections done based on previous comments.

Draft link:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

For Review.

Regards,
Ramkumar

**Comment by dmuelle** [comment]
a few comments-

in the bulleted list of files, start each list item with a capital letter

Replace these values according to the package that you want to verify.
--->
Replace the signature file and package version values according to the package that you want to verify.

They'll use the the same public key value so need to specify that only sig and version values are replaced.

To verify Liberty packages on Maven Central, download the public key to local machine. This example uses 46102B8E as the key id for the WebSphere Liberty public key.
--->
To verify Open Liberty packages on Maven Central, download the public key to your local machine by using [the gpg or gpg2 command](https://gnupg.org/). The key ID value for the WebSphere Liberty public key that you use to verify an Open Liberty package is `46102B8E`.


Edit the trust of the key owner. There are 5 levels of trust. 1. Do not know 2. Do not trust 3. Trust Marginally 4. Trust Fully 5. Trust Ultimately

Set trust at 5

--->

Use the `gpg2 --edit-key` command to [set the trust for the key owner](https://www.gnupg.org/gph/en/manual/x334.html) to 5. 

When you run the `gpg2 --edit-key` command, the console displays information about the key owner, followed by an internal command prompt. At this prompt, enter the `trust` command. You are then prompted to select a level of trust between 1 and 5. Enter 5.

Verify the file
--->
Verify the file by running the following command, This example verifies the Open Liberty 22.0.0.1 release package.

**Comment by dmuelle** [comment]
@cbridgha - a few more questions about he Maven info
- do you not need to download the zip or sig files, because they are in maven central and this method verifies them there?
- 
- **Chuck** For Maven Central - need to download zips and *.asc files
- 
- what is `openliberty-runtime-22.0.0.1.zip.asc`? Is this the package or the signature file? Do you not need a .sig file when verifying from Maven? 
- 
- **Chuck** The *.asc file is the signature file for Maven Central(*.sig files are not used here)
- 
- can you provide a more accurate/complete console example for the verifying trust step? I tried to run the gpg command to get and edit the key myself to get an example to copy but I get a `gpg: keyserver receive failed: Network is unreachable` message when I try to download the key. Not sure why, probably something I'm doing wrong or a prereq I don't have 
- 
- **Chuck**.  Yes this looks like a recent issue with MACOS - and you may need to install a more recent edition of pgp like here - https://sourceforge.net/p/gpgosx/docu/Download/.    It worked for me after upgrae - AND I was able to use the simplified command (minus the url prefix).  `gpg2 --keyserver keyserver.ubuntu.com --recv-keys 46102B8E`

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Review comments incorporated.

Draft link:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

For Review.

Regards,
Ramkumar

**Comment by dmuelle** [comment]
# Peer review

Per conversations with @NottyCode and @cbridgha, we should remove references to IBM from the text and avoid linking directly to the public key from this page. Link to the public key will be provided on the OL downloads page. Also, `.sig` files are only for verifiying with the openSSL command. Maven verification uses `.asc` files

Note that none of the links in the current draft work- probably some kind of syntax error

- Signature files (*.sig) are produced for every package of an Open Liberty release. You can verify these signature files by using the corresponding public key that is provided on the IBM DHE download site.
---->
Signature files are produced for every package of an Open Liberty release. You can use these signature files and the corresponding public key to verify the authenticity and integrity of an Open Liberty release package.

- IBM uses its private key to digitally sign the latest Open Liberty release. You can use the Open Liberty public key to check the signature, verify that the Open Liberty release package was released by IBM, and that it was not modified since its release.
--->
The Open Liberty project uses its private key to digitally sign each Open Liberty release. You can use the Open Liberty public key to check the signature, verify that the package was released by Open Liberty, and that it was not modified since its release.

- Remove the bulleted list and add:
"You can verify a release package either locally, by using the `openssl` command and a `*.sig` file, or on Maven Central, by using the `gpg` command."

Add an H2 heading for **Verifying Open Liberty packages with OpenSSL**  and the following sentence

To verify an Open Liberty release package locally,  you must first download an Open Liberty `.zip` package, the corresponding `.sig` file, and the Open LIberty public key. You can then run the `openssl` command to verify the package.

1. Go to the  **Downloads** section of the Open Liberty [Get Started page](https://www.openliberty.io/start/) and download an Open LIberty `.zip` package and its corresponding `.sig` file.
2. Obtain the public key file by using the public key link on the Get Started page.
3. After you download the files, you can run the following [OpenSSL command](https://www.openssl.org/docs/man1.1.1/man1/openssl-dgst.html) from the command line to verify the package. 
(example, explanation, and output example from the current draft)

## Verify Liberty packages on Maven Central
To verify Open Liberty packages on Maven Central, you must first download the public key to your local machine [the gpg or gpg2 command](https://gnupg.org/). You must edit the trust level for the key owner. You can then use this key to remotely verify an Open Liberty `.asc` release file on Maven Central. 

1. Run the following command to download the public key file. The key ID value for the public key is `46102B8E`.
(example)
2. Use the `gpg2 --edit-key` command to [set the trust for the key owner](https://www.gnupg.org/gph/en/manual/x334.html) to 5. When you run the gpg2 --edit-key command, the console displays information about the key owner, followed by an internal command prompt. At this prompt, enter the trust command. You are then prompted to select a level of trust. Enter 5.
(example- we need an updated example)
3. Verify the file by running the following command. This example verifies the Open Liberty 22.0.0.1 release package.
(example)
If the verification is successful, the command produces output that is similar to the following example.
(example)

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle - For Review.

Reworked on the file as suggested.

Draft link:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html


Regards,
Ramkumar

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle - For Review.

Reworked on the file as suggested.

Draft link:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

Regards,
Ramkumar

**Comment by dmuelle** [comment]
Looks good, a few comments

- there is still a problem with the link syntax in this topic- none of the links work
- in each subsection, replace the unordered steps with numbered steps
- in the intro, add the following sentence to the second paragraph:
"Signature files are available for Open Liberty releases in version 22.0.0.1 and later"
- in the first gpg2 command example, remove the `>` before the command
- "Verify the file by running the following command. This example verifies the Open Liberty 22.0.0.1 release package."
---->
Verify the file by running the `gpg2 --verify` command. The following example verifies the Open Liberty 22.0.0.1 release package.
- "If the verification is successful, the command produces the following output."
----> 
If the verification is successful, the command produces console output that is similar to the following example.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle - For Review.

Corrections done on the file as suggested.

Draft link:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

Regards,
Ramkumar

**Comment by dmuelle** [comment]
Hi @cbridgha -  I've been trying to follow the Maven process to generate an example of the console response for the "edit trust of key owner" step. But I can't get the `gpg2 --keyserver hkp://keyserver.ubuntu.com --recv-keys 46102B8E` command to work. Per your suggestion, I tried updating gpg- no dice. With both the LTS version (2.2.33) and the latest stable version (2.3.4) I get the same result: `gpg: keyserver receive failed: Network is unreachable`

3 questions-
1. since you were able to download the key, could you run `gpg2 --edit-key "WebSphere Liberty"` edit the trust to 5 and share the console output with us? I'd like to have an accurate example for that step since it  involves a secondary command prompt that is confusing to describe without an example.
2. When I updated from the LTS version to v 2.3.4, the command switched from gpg2 to gpg without any explanation. So after installing the latest stable release, gpg2 gives a command not found error and you have to use gpg instead :disappointed: . Should our examples use gpg or gpg2 ?
3.  Are the commands and process the same on WIndows for this step as they are on Mac/Linux?

**Comment by dmuelle** [comment]
console response for editing trust

```
$ gpg2 --edit-key "WebSphere Liberty"
gpg (GnuPG/MacGPG2) 2.2.32; Copyright (C) 2021 Free Software Foundation, Inc.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
pub rsa4096/4D210F6946102B8E
   created: 2021-06-02 expires: never    usage: SCEA
   trust: ultimate   validity: ultimate
[ultimate] (1). WebSphere Liberty (Works for IBM) <contbld@uk.ibm.com>
[ultimate] (2) WebSphereLiberty
gpg> trust
pub rsa4096/4D210F6946102B8E
   created: 2021-06-02 expires: never    usage: SCEA
   trust: ultimate   validity: ultimate
[ultimate] (1). WebSphere Liberty (Works for IBM) <contbld@uk.ibm.com>
[ultimate] (2) WebSphereLiberty
Please decide how far you trust this user to correctly verify other users' keys
(by looking at passports, checking fingerprints from different sources, etc.)
 1 = I don't know or won't say
 2 = I do NOT trust
 3 = I trust marginally
 4 = I trust fully
 5 = I trust ultimately
 m = back to the main menu
Your decision? 5
Do you really want to set this key to ultimate trust? (y/N) y
pub rsa4096/4D210F6946102B8E
   created: 2021-06-02 expires: never    usage: SCEA
   trust: ultimate   validity: ultimate
[ultimate] (1). WebSphere Liberty (Works for IBM) <contbld@uk.ibm.com>
[ultimate] (2) WebSphereLiberty
gpg>
```

**Comment by dmuelle** [comment]
I think we can break this up into multiple examples.

2. Use the `gpg2 --edit-key` command to set the trust for the key owner to 5.  

The following examples show the `gpg2 --edit-key` command and the console output that it generates.
```
gpg2 --edit-key "WebSphere Liberty"
gpg (GnuPG/MacGPG2) 2.2.32; Copyright (C) 2021 Free Software Foundation, Inc.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
pub rsa4096/4D210F6946102B8E
   created: 2021-06-02 expires: never    usage: SCEA
   trust: ultimate   validity: ultimate
[ultimate] (1). WebSphere Liberty (Works for IBM) <contbld@uk.ibm.com>
[ultimate] (2) WebSphereLiberty
```
When you run the `gpg2 --edit-key` command, the console displays information about the key owner, followed by an internal command prompt. At this prompt, enter the `trust` command, as shown in the following example. 

```
gpg> trust
pub rsa4096/4D210F6946102B8E
   created: 2021-06-02 expires: never    usage: SCEA
   trust: ultimate   validity: ultimate
[ultimate] (1). WebSphere Liberty (Works for IBM) <contbld@uk.ibm.com>
[ultimate] (2) WebSphereLiberty
```
You are then prompted to select a level of trust. Enter `5` and confirm your choice by entering `y`.

```
Please decide how far you trust this user to correctly verify other users' keys
(by looking at passports, checking fingerprints from different sources, etc.)
 1 = I don't know or won't say
 2 = I do NOT trust
 3 = I trust marginally
 4 = I trust fully
 5 = I trust ultimately
 m = back to the main menu
Your decision? 5
Do you really want to set this key to ultimate trust? (y/N) y
```

**Comment by dmuelle** [comment]
also noticed there is a stray plus sign before the second Results section

**Comment by ramkumar-k-9286** [comment]
Hi Chuck @cbridgha 

All the comments have been incorporated. The draft link for the page is:
https://draft-openlibertyio.mybluemix.net/docs/22.0.0.1/verifying-package-signatures.html

Please add the `Technical Reviewed` label if the page looks good.
Please add comments in the issue, in case you need any additional changes.

Regards,
Ramkumar

cc @dmuelle

**Comment by dmuelle** [comment]
doc is published with 22001: https://www.openliberty.io/docs/22.0.0.1/verifying-package-signatures.html

closing this issue
